docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
 parameter of the generic class ``typing.IO`` is constrained (only
 ``IO[str]``, ``IO[bytes]`` and ``IO[Any]`` are valid)::
 
   class IO(Generic[AnyStr]): ...
+
+
+Invalid overload definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should enforce the following rules for overload definitions.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+The ``@overload``-decorated definitions must be followed by an overload
+implementation, which does not include an ``@overload`` decorator. Type
+checkers should report an error or warning if an implementation is missing.
+Overload definitions within stub files, protocols, and abstract base classes
+are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Type checkers should report
+an error if these conditions are not met.
+
+If one or more overloads are decorated with ``@final`` or ``@override`` but the
+implementation is not, an error should be reported.
+
+Overloads are allowed to use a mixture of ``async def`` and ``def`` statements
+within the same overload definition. Type checkers should desugar all
+``async def`` statements before testing for implementation consistency
+and overlapping overloads (described below).
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that it is consistent with all of its associated overload signatures.
+The implementation should accept all potential sets of arguments
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be assignable to the input signatures
+of all overloads, and the return type of all overloads should be assignable to
+the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+When a type checker checks the implementation for consistency with overloads,
+it should first apply any transforms that change the effective type of the
+implementation including the presence of a ``yield`` statement in the
+implementation body, the use of ``async def``, and the presence of additional
+decorators.
+
+
+Overlapping overloads
+^^^^^^^^^^^^^^^^^^^^^
+
+If two overloads can accept the same set of arguments, they are said
```

Comment:
Hmm, that's a good point. This check would be pretty useless (i.e. way too noisy) if we were to interpret it in that way. Do you have any suggestions for how to restrict the rule? Maybe if we specifically say that "the potential of multiple inheritance should be ignored for purposes of this check". We'd presumably need to expand this to include all intersection types if/when they're added to the type system.

Reply:
Here's another example of overloads that mypy and pyright don't reject, but that do overlap in theory:

```
from typing import Protocol, overload

class HasFoo(Protocol):
    foo: int

class HasBar(Protocol):
    bar: str

@overload
def f(x: HasFoo) -> float: ...
@overload
def f(x: HasBar) -> bool: ...
def f(x: object) -> object:
    raise NotImplementedError
```

I don't have a good suggestion for an alternative way to phrase this rule.

Reply:
The purpose of the partial-overlap rule is to prevent unsoundness of this form:

```
@overload
def is_one(x: Literal[1]) -> True: ...
@overload
def is_one(x: int) -> False: ...

def f(x: int) -> bool:
    return is_one(x)

f(1)  # Presumably True at runtime, but type checker will say False (unless it performs inlining)
```

Where the runtime behavior may not match the inferred type of the call, because the type checker has correct but imprecise knowledge of the type of an argument.

The same unsoundness can occur with multiple inheritance:

```
@overload
def is_a(x: A) -> True: ...
@overload
def is_a(x: B) -> False: ...

class C(A, B): ...

def f(x: B) -> bool:
    return is_a(x)

f(C())  # True at runtime, type checker says False
```

And it can happen with overlapping structural types, as in Jelle's example: an object may actually conform to both protocols at runtime, but be statically typed as a less-precise type assignable only to the second protocol, resulting in incorrect inference.

I'm having trouble finding any principled reason why we shouldn't spec the rule as currently written here, since it seems correct. I'd be interested to see what actual typeshed overrides would fail this "presumed to be too noisy" version; it might illuminate what the real use cases are for overlapping overloads to violate this rule.

Reply:
Interesting, like Eric I had assumed that interpreting this rule strictly would be prohibitively noisy, but I looked over the overloads in typeshed's `builtins.pyi` just now, and the only clearly problematic case I could find was [`vars()`](https://github.com/python/typeshed/blob/f58dac1d62d33bb2255b762783d06463c40f5065/stdlib/builtins.pyi#L1759). This definition is indeed unsafe; here's an example of how it can be used to infer a wrong type: https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAsAFDkgCmAblQIYA2A%2BvAlQBQ30gDOHARgCUQyrQYs2nbnw6oYI8uQAmVYFGAcAHgC5CxITvJQTUanSatE0nvy2KKZVerTa9YAEYArKgGMFUAC0AHxQykj%2BANq8MCAANPpwALpGZKZmVDAAriAoUDJ2oo7mElbsHK7yIlAAxFC8ABZgAO68YREwCfT%2BWUyMcFAAsvQIyOgACuBacAAq1kA .

To describe the current state, we could add language saying that type checkers may use heuristics to skip reporting partial overlap in certain cases.

Reply:
I think the `vars` overload could be fixed to be compliant with this rule (and safe) by having the fallback overload return `Mapping`, which is assignable from `MappingProxyType`, instead of returning `dict`. But this wouldn't work if people are mutating the return value of `vars`. Which is only safe if they know the object passed in has a mutable `__dict__`, which the typing of `vars` can't verify; this unsoundness is explicit in the `Any` in the second overload.

It might make sense to extend the partial-overlap rule with a carve-out similar to step 5. That is, if overloads are partially overlapping, but some materialization of the second overload could make them not overlap at all, then we consider the possible unsoundness to have been explicitly requested, and we don't emit an overlap error.

Reply:
I've taken a crack at incorporating the above feedback in the latest draft. Let me know if you think additional changes are needed.

Reply:
I would personally prefer to specify only the "gradual guarantee" exemption based on materialization, and not the exemptions for intersections (via multiple inheritance or structural types), because I think this is more principled and sound.

However, I recognize that's a change from the current behavior of type checkers, and it probably needs more exploration and experimentation in real-world codebases than just an audit of overloads in builtins. So I'm ok with specifying the status quo for now (I'm happy to at least see these exemptions specified as a "may"), and adding this to my list of future changes to consider.